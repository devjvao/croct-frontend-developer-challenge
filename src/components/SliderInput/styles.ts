import styled, { css } from 'styled-components'

interface SliderProps {
  min: number
  max: number
  value: number
}

const getLinearGradientCss = (position: number) => css`
  linear-gradient(to right, #3f80ff 0%, #3f80ff ${position}%, #b9d1ff ${position}%, #b9d1ff 100%)
`

export const Slider = styled.input`
  border-radius: 0.1rem;
  height: 0.2rem;
  width: 100%;
  margin: 1.5rem 0;
  appearance: none;
  background: ${({ min, max, value }: SliderProps) =>
    getLinearGradientCss(((value - min) / (max - min)) * 100)};

  &::-ms-thumb,
  &::-moz-range-thumb,
  &::-webkit-slider-thumb {
    border: none;
    width: 1.2rem;
    height: 1.2rem;
    background: #3f80ff;
    border-radius: 50%;
  }
`
