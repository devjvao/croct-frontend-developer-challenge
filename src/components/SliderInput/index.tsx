import * as S from './styles'

interface Props {
  min: number
  max: number
  value: number
  onChange: (value: number) => void
}

const SliderInput = ({ min, max, value, onChange }: Props) => (
  <S.Slider
    type="range"
    min={min}
    max={max}
    value={value}
    onChange={({ target: { value: radius } }) => onChange(+radius)}
  />
)

export default SliderInput
