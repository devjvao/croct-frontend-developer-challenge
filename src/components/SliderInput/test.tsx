import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import SliderInput from '.'

describe('<SliderInput />', () => {
  it('should render the slider', () => {
    const { container } = render(
      <SliderInput min={100} max={300} value={200} onChange={jest.fn} />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should change slider value', () => {
    let value = 150

    const onChange = jest.fn((newValue: number) => {
      value = newValue
    })

    const { rerender } = render(
      <SliderInput min={100} max={300} value={value} onChange={onChange} />
    )
    const input = screen.getByRole('slider') as HTMLInputElement

    waitFor(() => {
      fireEvent.change(input, { target: { value: '151' } })
      fireEvent.change(input, { target: { value: '152' } })
      fireEvent.change(input, { target: { value: '153' } })
    })

    rerender(
      <SliderInput min={100} max={300} value={value} onChange={onChange} />
    )

    expect(onChange).toHaveBeenCalledTimes(3)
    expect(input.value).toBe('153')
  })
})
