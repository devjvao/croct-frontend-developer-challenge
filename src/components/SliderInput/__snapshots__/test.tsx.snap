// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SliderInput /> should render the slider 1`] = `
.c0 {
  border-radius: 0.1rem;
  height: 0.2rem;
  width: 100%;
  margin: 1.5rem 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: linear-gradient(to right,#3f80ff 0%,#3f80ff 50%,#b9d1ff 50%,#b9d1ff 100%);
}

.c0::-ms-thumb,
.c0::-moz-range-thumb,
.c0::-webkit-slider-thumb {
  border: none;
  width: 1.2rem;
  height: 1.2rem;
  background: #3f80ff;
  border-radius: 50%;
}

<input
  class="c0"
  max="300"
  min="100"
  role="slider"
  type="range"
  value="200"
/>
`;
